File: nodes/duration-calculator/duration-calculator.process.js
export default async ({inputs, settings, config, nodeConfig}) => {
  try {
    const startTime = inputs.start_time;
    const endTime = inputs.end_time || new Date().toISOString();
    const outputUnit = inputs.output_unit || 'seconds';
    
    // Parse timestamps
    const startDate = parseTimestamp(startTime);
    const endDate = parseTimestamp(endTime);
    
    if (!startDate || !endDate) {
      throw new Error("Invalid timestamp format. Use ISO strings, Unix timestamps, or parseable date formats");
    }
    
    // Calculate duration in milliseconds
    const durationMs = endDate.getTime() - startDate.getTime();
    const durationSeconds = durationMs / 1000;
    const durationMinutes = durationSeconds / 60;
    const durationHours = durationMinutes / 60;
    const durationDays = durationHours / 24;
    
    // Get duration in requested unit
    let duration;
    switch (outputUnit) {
      case 'seconds':
        duration = Math.round(durationSeconds * 100) / 100; // Round to 2 decimal places
        break;
      case 'minutes':
        duration = Math.round(durationMinutes * 100) / 100;
        break;
      case 'hours':
        duration = Math.round(durationHours * 100) / 100;
        break;
      case 'days':
        duration = Math.round(durationDays * 100) / 100;
        break;
      default:
        duration = Math.round(durationSeconds * 100) / 100;
    }
    
    // Generate human readable format
    const humanReadable = formatHumanReadable(durationMs);
    
    return {
      duration: duration,
      duration_seconds: Math.round(durationSeconds * 100) / 100,
      duration_minutes: Math.round(durationMinutes * 100) / 100,
      duration_hours: Math.round(durationHours * 100) / 100,
      duration_days: Math.round(durationDays * 100) / 100,
      human_readable: humanReadable
    };
    
  } catch (error) {
    throw new Error(`Duration Calculator error: ${error.message}`);
  }
};

// Helper function to parse various timestamp formats
function parseTimestamp(timestamp) {
  if (!timestamp) return null;
  
  // Try parsing as Unix timestamp (number)
  if (!isNaN(timestamp) && !isNaN(parseFloat(timestamp))) {
    const num = parseFloat(timestamp);
    // If it's a large number, assume it's milliseconds, otherwise seconds
    const date = num > 1000000000000 ? new Date(num) : new Date(num * 1000);
    if (!isNaN(date.getTime())) return date;
  }
  
  // Try parsing as ISO string or other date format
  const date = new Date(timestamp);
  if (!isNaN(date.getTime())) return date;
  
  return null;
}

// Helper function to format duration in human readable format
function formatHumanReadable(durationMs) {
  const seconds = Math.floor(durationMs / 1000);
  const minutes = Math.floor(seconds / 60);
  const hours = Math.floor(minutes / 60);
  const days = Math.floor(hours / 24);
  
  const remainingHours = hours % 24;
  const remainingMinutes = minutes % 60;
  const remainingSeconds = seconds % 60;
  
  const parts = [];
  
  if (days > 0) parts.push(`${days} day${days !== 1 ? 's' : ''}`);
  if (remainingHours > 0) parts.push(`${remainingHours} hour${remainingHours !== 1 ? 's' : ''}`);
  if (remainingMinutes > 0) parts.push(`${remainingMinutes} minute${remainingMinutes !== 1 ? 's' : ''}`);
  if (remainingSeconds > 0 || parts.length === 0) parts.push(`${remainingSeconds} second${remainingSeconds !== 1 ? 's' : ''}`);
  
  return parts.join(' ');
}

File: nodes/duration-calculator/duration-calculator.process.py
from datetime import datetime
from typing import Any, Dict, List, Union

async def process(inputs: Dict[str, Any], settings: Dict[str, Any], config: Dict[str, Any], nodeConfig: Dict[str, Any]) -> Dict[str, Any]:
    try:
        start_time = inputs.get('start_time')
        end_time = inputs.get('end_time') or datetime.now().isoformat() + 'Z'
        output_unit = inputs.get('output_unit', 'seconds')
        
        # Parse timestamps
        start_date = parse_timestamp(start_time)
        end_date = parse_timestamp(end_time)
        
        if not start_date or not end_date:
            raise Exception("Invalid timestamp format. Use ISO strings, Unix timestamps, or parseable date formats")
        
        # Calculate duration in seconds
        duration_seconds = (end_date - start_date).total_seconds()
        duration_minutes = duration_seconds / 60
        duration_hours = duration_minutes / 60
        duration_days = duration_hours / 24
        
        # Get duration in requested unit
        if output_unit == 'seconds':
            duration = round(duration_seconds, 2)
        elif output_unit == 'minutes':
            duration = round(duration_minutes, 2)
        elif output_unit == 'hours':
            duration = round(duration_hours, 2)
        elif output_unit == 'days':
            duration = round(duration_days, 2)
        else:
            duration = round(duration_seconds, 2)
        
        # Generate human readable format
        human_readable = format_human_readable(duration_seconds)
        
        return {
            'duration': duration,
            'duration_seconds': round(duration_seconds, 2),
            'duration_minutes': round(duration_minutes, 2),
            'duration_hours': round(duration_hours, 2),
            'duration_days': round(duration_days, 2),
            'human_readable': human_readable
        }
        
    except Exception as error:
        raise Exception(f"Duration Calculator error: {str(error)}")

def parse_timestamp(timestamp):
    """Helper function to parse various timestamp formats"""
    if not timestamp:
        return None
    
    try:
        # Try parsing as Unix timestamp (number)
        if isinstance(timestamp, (int, float)) or (isinstance(timestamp, str) and timestamp.replace('.', '').replace('-', '').isdigit()):
            num = float(timestamp)
            # If it's a large number, assume it's seconds, otherwise it might be milliseconds
            if num > 1000000000000:  # milliseconds
                return datetime.fromtimestamp(num / 1000)
            else:  # seconds
                return datetime.fromtimestamp(num)
        
        # Try parsing as ISO string or other date format
        return datetime.fromisoformat(timestamp.replace('Z', '+00:00'))
    except:
        return None

def format_human_readable(duration_seconds):
    """Helper function to format duration in human readable format"""
    seconds = int(duration_seconds)
    minutes = seconds // 60
    hours = minutes // 60
    days = hours // 24
    
    remaining_hours = hours % 24
    remaining_minutes = minutes % 60
    remaining_seconds = seconds % 60
    
    parts = []
    
    if days > 0:
        parts.append(f"{days} day{'s' if days != 1 else ''}")
    if remaining_hours > 0:
        parts.append(f"{remaining_hours} hour{'s' if remaining_hours != 1 else ''}")
    if remaining_minutes > 0:
        parts.append(f"{remaining_minutes} minute{'s' if remaining_minutes != 1 else ''}")
    if remaining_seconds > 0 or not parts:
        parts.append(f"{remaining_seconds} second{'s' if remaining_seconds != 1 else ''}")
    
    return ' '.join(parts)
